File: rd-text/src/main/kotlin/com/jetbrains/rider/rdtext/impl/ot/RdOtBasedText.kt
package com.jetbrains.rider.rdtext.impl.ot

import com.jetbrains.rider.framework.IRdDynamic
import com.jetbrains.rider.framework.base.RdDelegateBase
import com.jetbrains.rider.framework.impl.RdProperty
import com.jetbrains.rider.rdtext.*
import com.jetbrains.rider.rdtext.impl.intrinsics.RdAssertion
import com.jetbrains.rider.rdtext.impl.intrinsics.RdChangeOrigin
import com.jetbrains.rider.rdtext.impl.ot.intrinsics.RdAck
import com.jetbrains.rider.rdtext.impl.ot.intrinsics.RdOtState
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.reactive.*
import com.jetbrains.rider.util.reflection.usingTrueFlag


open class RdOtBasedText(delegate: RdOtState, final override val isMaster: Boolean = true) : RdDelegateBase<RdOtState>(delegate), ITextBuffer {
    private var isComplexChange = false
    private val textChanged: IOptProperty<RdTextChange> = OptProperty()

    private var diff: MutableList<OtOperation> = mutableListOf()

    private val _historyChanged: ISignal<RdTextChange> = Signal()

    override val historyChanged: ISource<RdTextChange> get() = _historyChanged
    final override var bufferVersion: TextBufferVersion = TextBufferVersion.INIT_VERSION
        private set

    private val localOrigin: RdChangeOrigin = if (isMaster) RdChangeOrigin.Master else RdChangeOrigin.Slave

    init {
        (delegatedBy.operation as RdProperty<*>).slave()
    }

    override fun bind(lf: Lifetime, parent: IRdDynamic, name: String) {
        super.bind(lf, parent, name)

        delegatedBy.operation.adviseNotNull(lf, {
            if (it.origin != localOrigin) receiveOperation(it)
        })
        delegatedBy.ack.advise(lf) {
            if (it.origin != localOrigin) updateHistory(it)
        }

        // for asserting documents equality
        delegatedBy.assertedSlaveText.compose(delegatedBy.assertedMasterText, { slave, master -> slave to master}).advise(lf) { (s, m) ->
            if (s.masterVersion == m.masterVersion
                    && s.slaveVersion == m.slaveVersion
                    && s.text != m.text) {
                throw IllegalStateException("Master and Slave texts are different.\nMaster:\n$m\nSlave:\n$s")
            }
        }
    }

    private fun updateHistory(ack: RdAck) {
        val ts = ack.timestamp
        diff.removeIf { it.timestamp == ts }
    }

    protected open fun receiveOperation(operation: OtOperation) {
        val remoteOrigin = operation.origin
        val transformedOp = when (operation.kind) {
            OtOperationKind.Normal -> {
                val (op, newDiff) = diff.fold(operation to mutableListOf<OtOperation>(), { (transformedOp, transformedDiff), localChange ->
                    val (transformedLocalChange, newTransformedOp) = transform(localChange, transformedOp)
                    transformedDiff.add(transformedLocalChange)
                    return@fold newTransformedOp to transformedDiff
                })

                diff = newDiff
                op
            }
            OtOperationKind.Reset -> {
                diff = mutableListOf()
                operation
            }
        }

        bufferVersion = if (remoteOrigin == RdChangeOrigin.Master)
            bufferVersion.incrementMaster()
        else
            bufferVersion.incrementSlave()

        val timestamp = operation.timestamp
        assert(if (remoteOrigin == RdChangeOrigin.Master) bufferVersion.master == timestamp else bufferVersion.slave == timestamp)

        val changes = transformedOp.toRdTextChanges()

        for ((i, ch) in changes.withIndex()) {
            usingTrueFlag(RdOtBasedText::isComplexChange, i < changes.lastIndex) {
                _historyChanged.fire(ch)
                textChanged.set(ch)
            }
        }


        if (operation.kind == OtOperationKind.Normal)
            sendAck(timestamp)
    }

    protected open fun sendAck(timestamp: Int) {
        delegatedBy.ack.fire(RdAck(timestamp, localOrigin))
    }

    override fun advise(lifetime: Lifetime, handler: (RdTextChange) -> Unit) {
        require(delegatedBy.isBound)
        protocol.scheduler.assertThread()

        textChanged.advise(lifetime, handler)
    }

    override fun fire(value: RdTextChange) {
        require(delegatedBy.isBound || bufferVersion == TextBufferVersion.INIT_VERSION)
        if (delegatedBy.isBound) protocol.scheduler.assertThread()

        bufferVersion = if (isMaster) bufferVersion.incrementMaster()
        else bufferVersion.incrementSlave()

        val ts = getCurrentTs()
        val operation = value.toOperation(localOrigin, ts)

        when (operation.kind) {
            OtOperationKind.Normal -> diff.add(operation)
            OtOperationKind.Reset -> diff = mutableListOf()
        }

        _historyChanged.fire(value)
        sendOperation(operation)
    }

    protected fun getCurrentTs() = if (isMaster) bufferVersion.master else bufferVersion.slave

    protected open fun sendOperation(operation: OtOperation) {
        delegatedBy.operation.set(operation)
    }

    override fun reset(text: String) {
        fire(RdTextChange(RdTextChangeKind.Reset, 0, "", text, text.length))
    }

    override fun assertState(allText: String) {
        if (isComplexChange) return
        val assertion = RdAssertion(bufferVersion.master, bufferVersion.slave, allText)
        val assertedTextProp = if (isMaster) delegatedBy.assertedMasterText else delegatedBy.assertedSlaveText
        assertedTextProp.set(assertion)
    }

    fun playCaret(caretOffset: Int): Int {
        if (isMaster) return caretOffset

        var newCaretOffset = caretOffset
        for (op in diff) {
            var offset = 0
            for (change in op.changes) {
                when (change) {
                    is DeleteText ->
                        if (offset < newCaretOffset) {
                            newCaretOffset -= change.text.length
                        }
                    is InsertText ->
                        if (offset < newCaretOffset) {
                            newCaretOffset += change.text.length
                        } else if (offset == newCaretOffset && change.priority != OtChangePriority.AfterAllChanges) {
                            newCaretOffset += change.text.length
                        }
                    is Retain -> offset += change.offset
                }
            }
        }
        return newCaretOffset
    }
}

File: rd-net/RdFramework/Text/Impl/Ot/RdOtBasedText.cs
using System;
using System.Collections.Generic;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Rd.Text.Impl.Intrinsics;
using JetBrains.Rd.Text.Impl.Ot.Intrinsics;
using JetBrains.Rd.Text.Intrinsics;

namespace JetBrains.Rd.Text.Impl.Ot
{
  public class RdOtBasedText : RdDelegateBase<RdOtState>, ITextBuffer
  {
    public bool IsMaster { get; set; } = false;

    private bool myIsComplexChange;

    private readonly IViewableProperty<RdTextChange> myTextChanged;

    private List<OtOperation> myDiff;

    public TextBufferVersion BufferVersion { get; private set; }

    private RdChangeOrigin LocalOrigin => IsMaster ? RdChangeOrigin.Master : RdChangeOrigin.Slave;

    public RdOtBasedText() : this(new RdOtState())
    {
    }

    public RdOtBasedText(RdOtState @delegate) : base(@delegate)
    {
      myTextChanged = new ViewableProperty<RdTextChange>();
      myDiff = new List<OtOperation>();

      // disabling mastering, text buffer must resolve conflicts by itself
      ((RdProperty<OtOperation>) Delegate.Operation).IsMaster = false;
    }

    class MasterSlave
    {
      public RdAssertion Master { get; }
      public RdAssertion Slave { get; }

      public MasterSlave(RdAssertion master, RdAssertion slave)
      {
        Master = master;
        Slave = slave;
      }
    }
    
    public override void Bind(Lifetime lf, IRdDynamic parent, string name)
    {
      base.Bind(lf, parent, name);

      Delegate.Operation.AdviseNotNull(lf, op =>
      {
        if (op.Origin != LocalOrigin) ReceiveOperation(op);
      });

      Delegate.Ack.Advise(lf, ack =>
      {
        if (ack.Origin != LocalOrigin) UpdateHistory(ack);
      });

      Delegate.AssertedMasterText.Compose(lf, Delegate.AssertedSlaveText, (m, s) => new MasterSlave(m, s)).Advise(lf, ms =>
      {
        var m = ms.Master;
        var s = ms.Slave;
        if (m.MasterVersion == s.MasterVersion
            && m.SlaveVersion == s.SlaveVersion
            && m.Text != s.Text)
        {
          throw new Exception($"Master and Slave texts are different.\nMaster:\n{m.Text}\nSlave:\n{s.Text}");
        }
      });
    }

    private void UpdateHistory(RdAck ack)
    {
      var ts = ack.Timestamp;
      myDiff.RemoveAll(op => op.Timestamp == ts);
    }

    private void ReceiveOperation(OtOperation operation)
    {
      var remoteOrigin = operation.Origin;
      var transformedOp = operation;
      switch (operation.Kind)
      {
        case OtOperationKind.Normal:
        {
          var newDiff = new List<OtOperation>();
          foreach (var localChange in myDiff)
          {
            var result = OtFramework.Transform(localChange, transformedOp);
            newDiff.Add(result.NewLocalDiff);
            transformedOp = result.LocalizedApplyToDocument;
          }

          myDiff = newDiff;

          break;
        }
        case OtOperationKind.Reset:
        {
          myDiff = new List<OtOperation>();
          break;
        }
        default:
          throw new ArgumentOutOfRangeException();
      }

      BufferVersion = remoteOrigin == RdChangeOrigin.Master
        ? BufferVersion.IncrementMaster()
        : BufferVersion.IncrementSlave();

      var timestamp = operation.Timestamp;
      Assertion.Assert(remoteOrigin == RdChangeOrigin.Master
        ? BufferVersion.Master == timestamp
        : BufferVersion.Slave == timestamp, "operation.Timestamp == BufferVersion");

      var changes = transformedOp.ToRdTextChanges();

      for (var i = 0; i < changes.Count; i++)
      {
        using (new ComplexChangeCookie(this, i < changes.Count - 1))
        {
          var change = changes[i];
          myTextChanged.SetValue(change);
        }
      }

      if (operation.Kind == OtOperationKind.Normal)
        SendAck(timestamp);
    }

    private void SendAck(int timestamp)
    {
      Delegate.Ack.Fire(new RdAck(timestamp, LocalOrigin));
    }

    public IScheduler Scheduler { get; set; }

    public void Fire(RdTextChange value)
    {
      Assertion.Assert(Delegate.IsBound || BufferVersion == TextBufferVersion.InitVersion, "Delegate.IsBound || BufferVersion == TextBufferVersion.InitVersion");
      if (Delegate.IsBound) Proto.Scheduler.AssertThread();

      BufferVersion = IsMaster ? BufferVersion.IncrementMaster() : BufferVersion.IncrementSlave();
      var ts = GetCurrentTs();
      var operation = value.ToOperation(LocalOrigin, ts);

      switch (operation.Kind)
      {
        case OtOperationKind.Normal:
          myDiff.Add(operation);
          break;
        case OtOperationKind.Reset:
          myDiff = new List<OtOperation>();
          break;
        default:
          throw new ArgumentOutOfRangeException();
      }

      SendOperation(operation);
    }

    private void SendOperation(OtOperation operation)
    {
      Delegate.Operation.SetValue(operation);
    }

    private int GetCurrentTs() => IsMaster ? BufferVersion.Master : BufferVersion.Slave;

    public void Advise(Lifetime lifetime, Action<RdTextChange> handler)
    {
      Assertion.Assert(Delegate.IsBound, "Delegate.IsBound");
      Proto.Scheduler.AssertThread();

      myTextChanged.Advise(lifetime, handler);
    }

    public void Reset(string text)
    {
      Fire(new RdTextChange(RdTextChangeKind.Reset, 0, "", text, text.Length));
    }

    public void AssertState(string allText)
    {
      if (myIsComplexChange) return;
      var assertion = new RdAssertion(BufferVersion.Master, BufferVersion.Slave, allText);
      if (IsMaster)
        Delegate.AssertedMasterText.SetValue(assertion);
      else
        Delegate.AssertedSlaveText.SetValue(assertion);
    }

    private struct ComplexChangeCookie : IDisposable
    {
      private readonly RdOtBasedText myOwner;
      private readonly bool myOldValue;

      public ComplexChangeCookie(RdOtBasedText owner, bool changeFlag = true)
      {
        myOwner = owner;
        myOldValue = owner.myIsComplexChange;
        if (changeFlag)
          owner.myIsComplexChange = true;
      }

      public void Dispose()
      {
        myOwner.myIsComplexChange = myOldValue;
      }
    }
  }
}